Incorrect Logic Results in Inability to Properly Configure Signature Algorithm

## problem

Incorrect Logic Results in Inability to Properly Configure Signature Algorithm

## detail

1. This is my application.yml file

   ```yml
   smallrye:
     jwt:
       sign:
         key: my-key
       new-token:
         signature-algorithm: ES256
         issuer: NJUPT
   ```

   You can see that I specified the encryption algorithm as ES256. It seemingly correct, right?

2. However, when I called the function ``` JwtClaimsBuilder.sign()```, program threw a exception

   ```text
   Caused by: java.lang.IllegalArgumentException: SRJWT05028: Signing key can not be created from the loaded content
   	at io.smallrye.jwt.build.impl.JwtSignatureImpl.sign(JwtSignatureImpl.java:101)
   ```

3. When this function was called in first time, it needed to read my private key by calling a method ``` getSigningKeyFromKeyContent() ```(marked with `here`)in ```JwtSignatureImpl```

   ```java
   public String sign() throws JwtSignatureException {
           try {
               Key key = configuredPemKey;
               if (key == null) {
                   String keyLocation = JwtBuildUtils.getConfigProperty(JwtBuildUtils.SIGN_KEY_LOCATION_PROPERTY, String.class);
                   if (keyLocation != null) {
                       key = JwtBuildUtils.readPrivateKeyFromKeystore(keyLocation.trim());
                       if (key == null) {
                           InputStream is = ResourceUtils.getResourceStream(keyLocation.trim());
                           if (is != null) {
                               try (InputStream keyStream = is) {
                                   key = getSigningKeyFromKeyContent(new String(ResourceUtils.readBytes(keyStream)));
                               }
                           }
                       }
                   } else {
                       key = JwtBuildUtils.readPrivateKeyFromKeystore(null);
                       if (key == null) {
                           String keyContent = JwtBuildUtils.getConfigProperty(JwtBuildUtils.SIGN_KEY_PROPERTY, String.class);
                           if (keyContent != null) {
                               key = getSigningKeyFromKeyContent(keyContent); //here
                           } else {
                               throw ImplMessages.msg.signKeyNotConfigured();
                           }
                       }
                   }
               }
               if (key == null) {
                   throw ImplMessages.msg.signingKeyCanNotBeCreatedFromContent();
               }
   
               return signInternal(key); //here2
           } catch (JwtSignatureException ex) {
               throw ex;
           } catch (Exception ex) {
               throw ImplMessages.msg.signatureException(ex);
           } finally {
               removeJti();
           }
       }
   ```

   ```java
   Key getSigningKeyFromKeyContent(String keyContent) {
           String kid = (String) headers.get("kid");
           String algHeader = (String) headers.get("alg"); //why?
   
           // Try PEM format first - default to RS256 if no algorithm header is set
           Key key = KeyUtils.tryAsPemSigningPrivateKey(keyContent,
                   (algHeader == null ? SignatureAlgorithm.RS256 : SignatureAlgorithm.fromAlgorithm(algHeader)));
           if (key == null) {
               if (kid == null) {
                   kid = JwtBuildUtils.getConfigProperty(JwtBuildUtils.SIGN_KEY_ID_PROPERTY, String.class);
                   if (kid != null) {
                       headers.put("kid", kid);
                   }
               }
   
               // Try to load JWK from a single JWK resource or JWK set resource
               JsonWebKey jwk = KeyUtils.getJwkKeyFromJwkSet(kid, keyContent);
               if (jwk != null) {
                   // if the user has already set the algorithm header then JWK `alg` header, if set, must match it
                   key = KeyUtils.getPrivateOrSecretSigningKey(jwk,
                           (algHeader == null ? null : SignatureAlgorithm.fromAlgorithm(algHeader)));
                   if (key != null) {
                       // if the algorithm header is not set then use JWK `alg`
                       if (algHeader == null && jwk.getAlgorithm() != null) {
                           headers.put("alg", jwk.getAlgorithm());
                       }
                       // if 'kid' header is not set then use JWK `kid`
                       if (kid == null && jwk.getKeyId() != null) {
                           headers.put("kid", jwk.getKeyId());
                       }
                   }
               }
           } else {
               configuredPemKey = key;
           }
           return key;
       }
   ```

   There is a bug here; the value ```algHeader``` obtained from the header has not been initialized.  So who will initialize it? Nobody.

4. But, there is a method could get signature algorithm.

   ```java
   private String getSignatureAlgorithm(Key signingKey) {
           String alg = (String) headers.get("alg");
           if (alg == null) {
               try {
                   alg = JwtBuildUtils.getConfigProperty(JwtBuildUtils.NEW_TOKEN_SIGNATURE_ALG_PROPERTY, String.class);
                   if (alg != null) {
                       alg = SignatureAlgorithm.valueOf(alg.toUpperCase()).getAlgorithm();
                   }
               } catch (Exception ex) {
                   throw ImplMessages.msg.unsupportedSignatureAlgorithm(alg);
               }
           }
           if ("none".equals(alg)) {
               throw ImplMessages.msg.noneSignatureAlgorithmUnsupported();
           }
           if (signingKey instanceof RSAPrivateKey) {
               if (alg == null) {
                   return SignatureAlgorithm.RS256.name();
               } else if (alg.startsWith("RS") || alg.startsWith("PS")) {
                   return alg;
               }
           } else if (signingKey instanceof ECPrivateKey) {
               if (alg == null) {
                   return SignatureAlgorithm.ES256.name();
               } else if (alg.startsWith("ES")) {
                   return alg;
               }
           } else if (signingKey instanceof SecretKey) {
               if (alg == null) {
                   return SignatureAlgorithm.HS256.name();
               } else if (alg.startsWith("HS")) {
                   return alg;
               }
           } else if (signingKey instanceof PrivateKey) {
               // for example, sun.security.pkcs11.P11Key$P11PrivateKey
               if (isEdECPrivateKey(signingKey)) {
                   if (alg == null || alg.equals(SignatureAlgorithm.EDDSA.getAlgorithm())) {
                       return SignatureAlgorithm.EDDSA.getAlgorithm();
                   }
               }
               if (alg == null) {
                   return SignatureAlgorithm.RS256.name();
               } else if (alg.startsWith("RS") || alg.startsWith("PS") || alg.startsWith("ES")) {
                   return alg;
               }
           }
           throw ImplMessages.msg.unsupportedSignatureAlgorithm(signingKey.getAlgorithm());
       }
   ```

   So which method calls it? A method called ```signInternal()```(marked with `here2` in the 3rd block),  which is used in function ```sign()```. 

   ```java
   private String signInternal(Key signingKey) {
           if (signingKey == null) {
               throw ImplMessages.msg.signingKeyIsNull();
           }
           JwtBuildUtils.setDefaultJwtClaims(claims, tokenLifespan);
           JsonWebSignature jws = new JsonWebSignature();
           for (Map.Entry<String, Object> entry : headers.entrySet()) {
               jws.setHeader(entry.getKey(), entry.getValue());
           }
           if (!headers.containsKey("typ")) {
               jws.setHeader("typ", "JWT");
           }
   
           String algorithm = getSignatureAlgorithm(signingKey);
   
           jws.setAlgorithmHeaderValue(algorithm);
   
           jws.setPayload(claims.toJson());
           jws.setKey(signingKey);
           if (isRelaxKeyValidation()) {
               jws.setDoKeyValidation(false);
           }
           try {
               return jws.getCompactSerialization();
           } catch (Exception ex) {
               throw ImplMessages.msg.signJwtTokenFailed(ex.getMessage(), ex);
           }
       }
   ```

5. I don't know why we need to get the signature algorithm at the same time as I get the key. It is helpless and makes a exception, and we can get the signature algorithm later in the process.

--- a/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtClaimsBuilderImpl.java
+++ b/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtClaimsBuilderImpl.java
@@ -202,7 +202,7 @@ class JwtClaimsBuilderImpl extends JwtSignatureImpl implements JwtClaimsBuilder,
      */
     @Override
     public JwtSignatureBuilder header(String name, Object value) {
-        if ("alg".equals(name)) {
+        if (HeaderParameterNames.ALGORITHM.equals(name)) {
             return algorithm(toSignatureAlgorithm((String) value));
         } else {
             headers.put(name, value);

--- a/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtEncryptionImpl.java
+++ b/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtEncryptionImpl.java
@@ -13,6 +13,7 @@ import javax.crypto.SecretKey;
 import org.jose4j.jwa.AlgorithmConstraints;
 import org.jose4j.jwe.JsonWebEncryption;
 import org.jose4j.jwk.JsonWebKey;
+import org.jose4j.jwx.HeaderParameterNames;
 
 import io.smallrye.jwt.algorithm.ContentEncryptionAlgorithm;
 import io.smallrye.jwt.algorithm.KeyEncryptionAlgorithm;
@@ -124,9 +125,9 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
      */
     @Override
     public JwtEncryptionBuilder header(String name, Object value) {
-        if ("alg".equals(name)) {
+        if (HeaderParameterNames.ALGORITHM.equals(name)) {
             return keyAlgorithm(toKeyEncryptionAlgorithm((String) value));
-        } else if ("enc".equals(name)) {
+        } else if (HeaderParameterNames.ENCRYPTION_METHOD.equals(name)) {
             return contentAlgorithm(toContentEncryptionAlgorithm((String) value));
         } else {
             headers.put(name, value);
@@ -139,7 +140,7 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
      */
     @Override
     public JwtEncryptionBuilder keyAlgorithm(KeyEncryptionAlgorithm algorithm) {
-        headers.put("alg", algorithm.getAlgorithm());
+        headers.put(HeaderParameterNames.ALGORITHM, algorithm.getAlgorithm());
         return this;
     }
 
@@ -148,7 +149,7 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
      */
     @Override
     public JwtEncryptionBuilder contentAlgorithm(ContentEncryptionAlgorithm algorithm) {
-        headers.put("enc", algorithm.getAlgorithm());
+        headers.put(HeaderParameterNames.ENCRYPTION_METHOD, algorithm.getAlgorithm());
         return this;
     }
 
@@ -157,7 +158,7 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
      */
     @Override
     public JwtEncryptionBuilder keyId(String keyId) {
-        headers.put("kid", keyId);
+        headers.put(HeaderParameterNames.KEY_ID, keyId);
         return this;
     }
 
@@ -171,8 +172,8 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
         for (Map.Entry<String, Object> entry : headers.entrySet()) {
             jwe.getHeaders().setObjectHeaderValue(entry.getKey(), entry.getValue());
         }
-        if (innerSigned && !headers.containsKey("cty")) {
-            jwe.getHeaders().setObjectHeaderValue("cty", "JWT");
+        if (innerSigned && !headers.containsKey(HeaderParameterNames.CONTENT_TYPE)) {
+            jwe.getHeaders().setObjectHeaderValue(HeaderParameterNames.CONTENT_TYPE, "JWT");
         }
         String keyAlgorithm = getKeyEncryptionAlgorithm(key);
         jwe.setAlgorithmConstraints(new AlgorithmConstraints(AlgorithmConstraints.ConstraintType.PERMIT, keyAlgorithm));
@@ -193,18 +194,24 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
         return JwtBuildUtils.getConfigProperty(JwtBuildUtils.ENC_KEY_RELAX_VALIDATION_PROPERTY, Boolean.class, false);
     }
 
-    private String getKeyEncryptionAlgorithm(Key keyEncryptionKey) {
-        String alg = (String) headers.get("alg");
+    private String getConfiguredKeyEncryptionAlgorithm() {
+        String alg = (String) headers.get(HeaderParameterNames.ALGORITHM);
         if (alg == null) {
             try {
                 alg = JwtBuildUtils.getConfigProperty(JwtBuildUtils.NEW_TOKEN_KEY_ENCRYPTION_ALG_PROPERTY, String.class);
                 if (alg != null) {
                     alg = KeyEncryptionAlgorithm.fromAlgorithm(alg).getAlgorithm();
+                    headers.put(HeaderParameterNames.ALGORITHM, alg);
                 }
             } catch (Exception ex) {
                 throw ImplMessages.msg.unsupportedKeyEncryptionAlgorithm(alg);
             }
         }
+        return alg;
+    }
+
+    private String getKeyEncryptionAlgorithm(Key keyEncryptionKey) {
+        String alg = getConfiguredKeyEncryptionAlgorithm();
 
         if (keyEncryptionKey instanceof RSAPublicKey) {
             if (alg == null) {
@@ -233,7 +240,7 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
     }
 
     private String getContentEncryptionAlgorithm() {
-        String alg = (String) headers.get("enc");
+        String alg = (String) headers.get(HeaderParameterNames.ENCRYPTION_METHOD);
         if (alg == null) {
             try {
                 alg = JwtBuildUtils.getConfigProperty(JwtBuildUtils.NEW_TOKEN_CONTENT_ENCRYPTION_ALG_PROPERTY, String.class);
@@ -256,18 +263,17 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
     }
 
     Key getEncryptionKeyFromKeyContent(String keyContent) {
-        String kid = (String) headers.get("kid");
-        String algHeader = (String) headers.get("alg");
+        String kid = (String) headers.get(HeaderParameterNames.KEY_ID);
+        String alg = getConfiguredKeyEncryptionAlgorithm();
 
         // Try PEM format first - default to RSA_OAEP_256 if no algorithm header is set
         Key key = KeyUtils.tryAsPemEncryptionPublicKey(keyContent,
-                (algHeader == null ? KeyEncryptionAlgorithm.RSA_OAEP_256
-                        : KeyEncryptionAlgorithm.fromAlgorithm(algHeader)));
+                (alg == null ? KeyEncryptionAlgorithm.RSA_OAEP_256 : KeyEncryptionAlgorithm.fromAlgorithm(alg)));
         if (key == null) {
             if (kid == null) {
                 kid = JwtBuildUtils.getConfigProperty(JwtBuildUtils.ENC_KEY_ID_PROPERTY, String.class);
                 if (kid != null) {
-                    headers.put("kid", kid);
+                    headers.put(HeaderParameterNames.KEY_ID, kid);
                 }
             }
             // Try to load JWK from a single JWK resource or JWK set resource
@@ -275,15 +281,15 @@ class JwtEncryptionImpl implements JwtEncryptionBuilder {
             if (jwk != null) {
                 // if the user has already set the algorithm header then JWK `alg` header, if set, must match it
                 key = KeyUtils.getPublicOrSecretEncryptingKey(jwk,
-                        (algHeader == null ? null : KeyEncryptionAlgorithm.fromAlgorithm(algHeader)));
+                        (alg == null ? null : KeyEncryptionAlgorithm.fromAlgorithm(alg)));
                 if (key != null) {
                     // if the algorithm header is not set then use JWK `alg`
-                    if (algHeader == null && jwk.getAlgorithm() != null) {
-                        headers.put("alg", jwk.getAlgorithm());
+                    if (alg == null && jwk.getAlgorithm() != null) {
+                        headers.put(HeaderParameterNames.ALGORITHM, jwk.getAlgorithm());
                     }
                     // if 'kid' header is not set then use JWK `kid`
                     if (kid == null && jwk.getKeyId() != null) {
-                        headers.put("kid", jwk.getKeyId());
+                        headers.put(HeaderParameterNames.KEY_ID, jwk.getKeyId());
                     }
                 }
             }

--- a/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtSignatureImpl.java
+++ b/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtSignatureImpl.java
@@ -14,6 +14,7 @@ import org.eclipse.microprofile.jwt.Claims;
 import org.jose4j.jwk.JsonWebKey;
 import org.jose4j.jws.JsonWebSignature;
 import org.jose4j.jwt.JwtClaims;
+import org.jose4j.jwx.HeaderParameterNames;
 
 import io.smallrye.jwt.algorithm.SignatureAlgorithm;
 import io.smallrye.jwt.build.JwtEncryptionBuilder;
@@ -169,8 +170,8 @@ class JwtSignatureImpl implements JwtSignature {
         for (Map.Entry<String, Object> entry : headers.entrySet()) {
             jws.setHeader(entry.getKey(), entry.getValue());
         }
-        if (!headers.containsKey("typ")) {
-            jws.setHeader("typ", "JWT");
+        if (!headers.containsKey(HeaderParameterNames.TYPE)) {
+            jws.setHeader(HeaderParameterNames.TYPE, "JWT");
         }
 
         String algorithm = getSignatureAlgorithm(signingKey);
@@ -193,18 +194,24 @@ class JwtSignatureImpl implements JwtSignature {
         return JwtBuildUtils.getConfigProperty(JwtBuildUtils.SIGN_KEY_RELAX_VALIDATION_PROPERTY, Boolean.class, false);
     }
 
-    private String getSignatureAlgorithm(Key signingKey) {
-        String alg = (String) headers.get("alg");
+    private String getConfiguredSignatureAlgorithm() {
+        String alg = (String) headers.get(HeaderParameterNames.ALGORITHM);
         if (alg == null) {
             try {
                 alg = JwtBuildUtils.getConfigProperty(JwtBuildUtils.NEW_TOKEN_SIGNATURE_ALG_PROPERTY, String.class);
                 if (alg != null) {
                     alg = SignatureAlgorithm.valueOf(alg.toUpperCase()).getAlgorithm();
+                    headers.put(HeaderParameterNames.ALGORITHM, alg);
                 }
             } catch (Exception ex) {
                 throw ImplMessages.msg.unsupportedSignatureAlgorithm(alg);
             }
         }
+        return alg;
+    }
+
+    private String getSignatureAlgorithm(Key signingKey) {
+        String alg = getConfiguredSignatureAlgorithm();
         if ("none".equals(alg)) {
             throw ImplMessages.msg.noneSignatureAlgorithmUnsupported();
         }
@@ -255,17 +262,17 @@ class JwtSignatureImpl implements JwtSignature {
     }
 
     Key getSigningKeyFromKeyContent(String keyContent) {
-        String kid = (String) headers.get("kid");
-        String algHeader = (String) headers.get("alg");
+        String kid = (String) headers.get(HeaderParameterNames.KEY_ID);
+        String alg = getConfiguredSignatureAlgorithm();
 
-        // Try PEM format first - default to RS256 if no algorithm header is set
+        // Try PEM format first - default to RS256 if the algorithm is unknown
         Key key = KeyUtils.tryAsPemSigningPrivateKey(keyContent,
-                (algHeader == null ? SignatureAlgorithm.RS256 : SignatureAlgorithm.fromAlgorithm(algHeader)));
+                (alg == null ? SignatureAlgorithm.RS256 : SignatureAlgorithm.fromAlgorithm(alg)));
         if (key == null) {
             if (kid == null) {
                 kid = JwtBuildUtils.getConfigProperty(JwtBuildUtils.SIGN_KEY_ID_PROPERTY, String.class);
                 if (kid != null) {
-                    headers.put("kid", kid);
+                    headers.put(HeaderParameterNames.KEY_ID, kid);
                 }
             }
 
@@ -274,15 +281,15 @@ class JwtSignatureImpl implements JwtSignature {
             if (jwk != null) {
                 // if the user has already set the algorithm header then JWK `alg` header, if set, must match it
                 key = KeyUtils.getPrivateOrSecretSigningKey(jwk,
-                        (algHeader == null ? null : SignatureAlgorithm.fromAlgorithm(algHeader)));
+                        (alg == null ? null : SignatureAlgorithm.fromAlgorithm(alg)));
                 if (key != null) {
                     // if the algorithm header is not set then use JWK `alg`
-                    if (algHeader == null && jwk.getAlgorithm() != null) {
-                        headers.put("alg", jwk.getAlgorithm());
+                    if (alg == null && jwk.getAlgorithm() != null) {
+                        headers.put(HeaderParameterNames.ALGORITHM, jwk.getAlgorithm());
                     }
                     // if 'kid' header is not set then use JWK `kid`
                     if (kid == null && jwk.getKeyId() != null) {
-                        headers.put("kid", jwk.getKeyId());
+                        headers.put(HeaderParameterNames.KEY_ID, jwk.getKeyId());
                     }
                 }
             }

--- a/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtBuildConfigSource.java
+++ b/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtBuildConfigSource.java
@@ -22,6 +22,8 @@ public class JwtBuildConfigSource implements ConfigSource {
             JwtBuildUtils.NEW_TOKEN_AUDIENCE_PROPERTY,
             JwtBuildUtils.NEW_TOKEN_LIFESPAN_PROPERTY,
             JwtBuildUtils.NEW_TOKEN_OVERRIDE_CLAIMS_PROPERTY,
+            JwtBuildUtils.NEW_TOKEN_SIGNATURE_ALG_PROPERTY,
+            JwtBuildUtils.NEW_TOKEN_KEY_ENCRYPTION_ALG_PROPERTY,
             JwtBuildUtils.SIGN_KEYSTORE_KEY_ALIAS,
             JwtBuildUtils.ENC_KEYSTORE_KEY_ALIAS));
 

--- a/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtEncryptTest.java
+++ b/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtEncryptTest.java
@@ -359,6 +359,32 @@ public class JwtEncryptTest {
     }
 
     @Test
+    void encryptWithConfiguredEcKeyAndAlgorithmAndA128CBCHS256() throws Exception {
+        JwtBuildConfigSource configSource = JwtSignTest.getConfigSource();
+        configSource.setEncryptionKeyLocation("/ecPublicKey.pem");
+        configSource.setKeyEncryptionAlgorithm("ECDH-ES+A256KW");
+        String jweCompact = null;
+        try {
+            jweCompact = Jwt.claims()
+                    .claim("customClaim", "custom-value")
+                    .jwe()
+                    .keyId("key-enc-key-id")
+                    .contentAlgorithm(ContentEncryptionAlgorithm.A128CBC_HS256)
+                    .encrypt();
+        } finally {
+            configSource.setEncryptionKeyLocation("/publicKey.pem");
+            configSource.setKeyEncryptionAlgorithm(null);
+        }
+
+        checkJweHeaders(jweCompact, "ECDH-ES+A256KW", "A128CBC-HS256", 4);
+
+        JsonWebEncryption jwe = getJsonWebEncryption(jweCompact, getEcPrivateKey());
+
+        JwtClaims claims = JwtClaims.parse(jwe.getPlaintextString());
+        checkJwtClaims(claims);
+    }
+
+    @Test
     void encryptWithConfiguredEcKeyAndContentAlgorithm() throws Exception {
         JwtBuildConfigSource configSource = JwtSignTest.getConfigSource();
         configSource.setEncryptionKeyLocation("/ecPublicKey.pem");

--- a/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtSignTest.java
+++ b/implementation/jwt-build/src/test/java/io/smallrye/jwt/build/JwtSignTest.java
@@ -892,6 +892,30 @@ class JwtSignTest {
         assertEquals("custom-value", claims.getClaimValue("customClaim"));
     }
 
+    @Test
+    void signClaimsEcKeyFileWithConfiguredAlgorithm() throws Exception {
+        JwtBuildConfigSource configSource = getConfigSource();
+        configSource.setSigningKeyLocation("/ecPrivateKey.pem");
+        configSource.setSignatureAlgorithm(SignatureAlgorithm.ES256.getAlgorithm());
+        String jwt = null;
+        try {
+            jwt = Jwt.claim("customClaim", "custom-value")
+                    .sign();
+        } finally {
+            configSource.setSigningKeyLocation("/privateKey.pem");
+            configSource.setSignatureAlgorithm(null);
+        }
+
+        PublicKey ecKey = getEcPublicKey();
+        JsonWebSignature jws = getVerifiedJws(jwt, ecKey);
+        JwtClaims claims = JwtClaims.parse(jws.getPayload());
+
+        assertEquals(4, claims.getClaimsMap().size());
+        Map<String, Object> headers = getJwsHeaders(jwt, 2);
+        checkDefaultClaimsAndHeaders(headers, claims, "ES256", 300);
+        assertEquals("custom-value", claims.getClaimValue("customClaim"));
+    }
+
     private static SecretKey createSecretKey() throws Exception {
         String jwkJson = "{\"kty\":\"oct\",\"k\":\"Fdh9u8rINxfivbrianbbVT1u232VQBZYKx1HGAGPt2I\"}";
         JsonWebKey jwk = JsonWebKey.Factory.newJwk(jwkJson);

